### PR TITLE
fix mypy typing errors in pytorch_lightning/strategies/strategy.py

### PR DESCRIPTION
**Pull request by Cyprien-Ricque**
## What does this PR do?

Fixes mypy typing errors in pytorch_lightning/strategies/strategy.py in https://github.com/Lightning-AI/lightning/issues/13445.

- ```model``` becomes explicitly ```Optional[]``` in plugins functions such as ```optimizer_step``` and ```_after_closure```
- ```self.accelerator``` becomes explicitly ```Optional[]``` in strategy classes
- The existence of ```self.accelerator```, ```self.model``` and ```self.lightning_module``` is checked in strategy classes using ```assert``` when needed
- Variable ```output``` in ```training_step_end```, ```validation_step_end``` and ```test_step_end``` has been renamed to ```step_output``` to stay consistent with other functions
- ```self.model``` has to be of type ```pl.LightningModule``` or ```_LightningPrecisionModuleWrapperBase``` to use ```training_step```, ```validation_step```, ```test_step``` and ```predict_step```

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: src/pytorch_lightning/strategies/strategy.py [hunk 1]
```diff
@@ -128,6 +129,7 @@ def setup_optimizers(self, trainer: "pl.Trainer") -> None:
         """
         if trainer.state.fn not in (TrainerFn.FITTING, TrainerFn.TUNING):
             return
+        assert self.lightning_module, "self.lightning_module must be set before _init_optimizers_and_lr_schedulers()"
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply src/pytorch_lightning/strategies/strategy.py:132]: Instead, let the code speak for itself:
```suggestion
        assert self.lightning_module is not None
```

File: src/pytorch_lightning/strategies/strategy.py [hunk 2]
```diff
@@ -327,49 +335,61 @@ def load_optimizer_state_dict(self, checkpoint: Mapping[str, Any]) -> None:
...
             return self.model.validation_step(*args, **kwargs)
 
-    def test_step(self, *args, **kwargs) -> Optional[STEP_OUTPUT]:
+    def test_step(self, *args: Any, **kwargs: Any) -> Optional[STEP_OUTPUT]:
         """The actual test step.
 
         See :meth:`~pytorch_lightning.core.module.LightningModule.test_step` for more details
         """
         with self.precision_plugin.test_step_context():
+            assert isinstance(self.model, pl.LightningModule) or isinstance(
```
Comment (awaelchli) [inline src/pytorch_lightning/strategies/strategy.py:369]: Hmm, not sure this would work with custom strategies who have their own wrappers. 
Shouldn't we check against a protocol here, e.g., a type expressing that the object has signatures for training_step, test_step, etc. to be able to call these methods? Essentially we want to rely on duck-typing here. Does that make sense?

File: src/pytorch_lightning/strategies/dp.py [hunk 3]
```diff
@@ -137,17 +137,17 @@ def predict_step(self, *args, **kwargs) -> STEP_OUTPUT:
         with self.precision_plugin.predict_step_context():
             return self.model(*args, **kwargs)
 
-    def training_step_end(self, output):
+    def training_step_end(self, step_output):
```
[the remark this answers is not in the document]
Comment (Cyprien-Ricque) [reply src/pytorch_lightning/strategies/dp.py:140]: Ok! I am undoing this

File: src/pytorch_lightning/plugins/precision/apex_amp.py [hunk 4]
```diff
@@ -76,7 +76,7 @@ def backward(
 
     def optimizer_step(
         self,
-        model: Union["pl.LightningModule", Module],
+        model: Optional[Union["pl.LightningModule", Module]],
```
[the remark this answers is not in the document]
Comment (Cyprien-Ricque) [reply src/pytorch_lightning/plugins/precision/apex_amp.py:79]: Because it inherits from ```MixedPrecisionPlugin``` which inherits from ```PrecisionPlugin```. ```PrecisionPlugin``` requires model to be Optional otherwise the 2 first configurations of [this test](https://github.com/Lightning-AI/lightning/blob/03cbca430d7d82756b2786022cddcf767d3ffb89/tests/tests_pytorch/lite/test_parity.py#L118) does not run.

File: src/pytorch_lightning/strategies/strategy.py (hunk 2 above)
Comment (Cyprien-Ricque) [reply src/pytorch_lightning/strategies/strategy.py]: I think the idea was to implement something like 
```python
@runtime_checkable
class Trainable(Protocol):
    """Trainable protocol requires all ``*_step`` so that it can safely be used by ``Strategy`` classes."""
    def training_step(self, *args: Any, **kwargs: Any) -> STEP_OUTPUT:
        ...

    def validation_step(self, *args: Any, **kwargs: Any) -> Optional[STEP_OUTPUT]:
        ...

    def test_step(self, *args: Any, **kwargs: Any) -> Optional[STEP_OUTPUT]:
        ...

    def predict_step(self, *args: Any, **kwargs: Any) -> STEP_OUTPUT:
        ...
```
However this is not flexible enough for the classes implementing only few of them such as [this test model](https://github.com/Lightning-AI/lightning/blob/61c28cb428a13c2aea6d7f3f55e0f00431a4ea4e/tests/tests_pytorch/trainer/test_dataloaders.py#L1418) which does not implement ```validation_step```.
Therefore I split those functions in 4 different protocols. Even though it works I'm not sure it's clean.

File: src/pytorch_lightning/plugins/precision/apex_amp.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (justusschock) [reply src/pytorch_lightning/plugins/precision/apex_amp.py:79]: I get the point of inheritance but not why the test would fail, the model there is always defined as a boringmodel?

[the remark this answers is not in the document]
Comment (awaelchli) [reply src/pytorch_lightning/plugins/precision/apex_amp.py:79]: Ok got it. It's because of Lite where the model is optional and only passed to optimizer_step in case of deepspeed. So in general it remains optional.

**Review by awaelchli — APPROVED** [review]
Great work. Allowed myself to make minor cosmetic edits :)

**Review by otaj — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]

File: src/pytorch_lightning/plugins/precision/deepspeed.py [hunk 5]
```diff
@@ -88,7 +88,7 @@ def _run_backward(
 
     def optimizer_step(
         self,
-        model: Union["pl.LightningModule", Module],
+        model: Optional[Union["pl.LightningModule", Module]],
```
Comment (Borda) [inline src/pytorch_lightning/plugins/precision/deepspeed.py:91]:
```suggestion
        model: Union["pl.LightningModule", Module, None],
```
Maybe, fewer brackets?

**Review by Borda — APPROVED** [review]

File: src/pytorch_lightning/plugins/precision/deepspeed.py (hunk 5 above)
Comment (otaj) [reply src/pytorch_lightning/plugins/precision/deepspeed.py:91]: I myself am fine with both and find it a matter of preference.

**Review by justusschock — APPROVED** [review]

File: src/pytorch_lightning/plugins/precision/deepspeed.py (hunk 5 above)
Comment (justusschock) [reply src/pytorch_lightning/plugins/precision/deepspeed.py:91]: I would stick with `Optional` as this is what we have in other places as well :)
